### PR TITLE
feat: enforcer reactjs test

### PR DESCRIPTION
**Pull request by Shivansh-yadav13**
- added ReactJS test of enforcer.
- test/reactjs/src/Enforcer.js

### CODE REVIEW

**Comment by Shivansh-yadav13** [comment]
@hsluoyz @Zxilly @nodece pls review

**Comment by hsluoyz** [comment]
@nodece plz review

**Comment by hsluoyz** [comment]
@Shivansh-yadav13 plz fix:

image

**Comment by nodece** [comment]
@Shivansh-yadav13 I switch commit into the beta branch, please fix the conflict.

**Comment by Zxilly** [comment]
You can make eslint ignore folder `test` to pass CI.

**Review by Zxilly — APPROVED** [review]
It will much better if clean up some start-up comment in the cra template.

**Comment by nodece** [comment]
I suggest that add a workflow to build react project to check compatibility, you can make a new PR to do this.

**Review by nodece — APPROVED** [review]

**Comment by nodece** [comment]
@Zxilly Could you help review this PR again?

File: test/reactjs/package.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  "private": true,
+  "dependencies": {
+    "@craco/craco": "^6.4.3",
+    "@testing-library/jest-dom": "^5.14.1",
+    "@testing-library/react": "^13.0.0",
+    "@testing-library/user-event": "^13.2.1",
+    "react": "^18.1.0",
+    "react-dom": "^18.1.0",
+    "react-scripts": "5.0.1",
+    "web-vitals": "^2.1.0"
```
Comment (Zxilly) [inline test/reactjs/package.json]: Maybe you should also add
```suggestion
    "web-vitals": "^2.1.0"
    "casbin-core": "file:../../"
```

**Review by Zxilly — CHANGES_REQUESTED** [review]
I failed to run the unit tests. Maybe need some edit?
image

File: test/reactjs/package.json (hunk 1 above)
Comment (Shivansh-yadav13) [reply test/reactjs/package.json]: @Zxilly do you know how I can fix `ReferenceError: await is not defined` 😅

Comment (Zxilly) [reply test/reactjs/package.json]: @Shivansh-yadav13 `top-level-await` only works in esm module.
I write it like this
```js
import React, { useEffect, useState } from "react";
import { MemoryAdapter, Model, newEnforcer } from "casbin-core";

const model = new Model(`
    [request_definition]
    r = sub, obj, act

    [policy_definition]
    p = sub, obj, act

    [policy_effect]
    e = some(where (p.eft == allow))

    [matchers]
    m = r.sub == p.sub && r.obj == p.obj && r.act == p.act
    `);

const adapter = new MemoryAdapter(
  `p, alice, data1, read
p, bob, data2, write`);

const Enforcer = ({ sub, obj, act }) => {
  const [loading, setLoading] = useState(false);
  const [result, setResult] = useState(false);
  const [enforcer, setEnforcer] = useState(null);


  useEffect(() => {
    console.log("Enforcer useEffect");
    newEnforcer(model, adapter).then(e => {
      setEnforcer(e);
    });
  }, []);

  useEffect(() => {
    if (enforcer) {
      setLoading(true);
      enforcer.enforce(sub, obj, act).then((res) => setResult(res)).finally(() => setLoading(false));
    }
  }, [sub, obj, act, enforcer]);
  return (
    <>
      <h1>Casbin</h1>
      <h2>Request ({sub}, {obj}, {act})</h2>
      {
        loading ?
          <h3>Enforcing...</h3>
          :
          <h3 data-testid="enforce-result">{result ? "Allowed" : "Denied"}</h3>
      }
    </>
  );
};

export default Enforcer;
```
Maybe can wrap enforcer initialization in `useCallback()`.
@nodece What's your suggestion?

Comment (nodece) [reply test/reactjs/package.json]: I remember support top-level await.

Comment (nodece) [reply test/reactjs/package.json]: @Zxilly This code looks great to me.

Comment (nodece) [reply test/reactjs/package.json]: > Maybe you should also add

Using  `"casbin-core": "link:../../"` for refer to casbin-core/lib.

**Review by nodece — APPROVED** [review]

**Comment by Shivansh-yadav13** [comment]
> I suggest that add a workflow to build react project to check compatibility, you can make a new PR to do this.

I will make PR for the workflow. Then we will merge it.

**Review by Zxilly — APPROVED** [review]
lgtm

**Pull request merged**